pinephone/vibrator/FFDevice.cpp

```diff
@@ -0,0 +1,149 @@
+    if (FFDevice::event_fd > 0) {
+        LOG(INFO) << "FFDevice already initialized";
+        return false;
+    }
+
+    /* Checks if path to event in /dev/input/ not loaded */
+    if (!FFDevice::path_loaded) {
+        if (!FFDevice::findEventPath(input_path, FFDevice::event_path)) {
+            return false;
+        }
+        FFDevice::path_loaded = true;
+        FFDevice::rumble_effect_id = -1;
+    }
+
+    FFDevice::event_fd = open(FFDevice::event_path, O_RDWR);
+    LOG(INFO) << "FFDevice: open event: " << FFDevice::event_path;
+    if (event_fd == -1) {
+        LOG(INFO) << "FFDevice: open event error: " << strerror(errno);
+        FFDevice::event_fd = 0;
+        return false;
+    }
+
+    /* Checks if an effect is loaded into the driver */
+    if (FFDevice::rumble_effect_id == -1) {
+        /* Prepare effect */
+        FFDevice::effect.type = FF_RUMBLE;
+	    FFDevice::effect.u.rumble.strong_magnitude = 65535;
+	    FFDevice::effect.u.rumble.weak_magnitude = 65535;
+	    FFDevice::effect.replay.length = 1000;
+	    FFDevice::effect.replay.delay = 0;
+	    FFDevice::effect.id = -1;
+
+        /* Send the effect to the driver */
+        if (ioctl(FFDevice::event_fd, EVIOCSFF, &FFDevice::effect) == -1) {
+            LOG(INFO) << "FFDevice: send effect failed";
+	    }
+        FFDevice::rumble_effect_id = effect.id;
+    
+	    /* Prepare the stop event */
+	    FFDevice::stop.type = EV_FF;
+	    FFDevice::stop.code = FFDevice::rumble_effect_id;
+	    FFDevice::stop.value = 0;
+
+	    /* Prepare the play event */
+	    FFDevice::play.type = EV_FF;
+	    FFDevice::play.code = FFDevice::rumble_effect_id;
+	    FFDevice::play.value = 1;
+    }
+    return true;
+}
+
+void FFDevice::vibrate(int duration_ms) {
+    if (FFDevice::event_fd == 0) {
+        return;
+    }
+
+	FFDevice::effect.replay.length = duration_ms;
+
+    /* Changing the effect in the driver */
+    if (ioctl(FFDevice::event_fd, EVIOCSFF, &FFDevice::effect) == -1) {
+        LOG(INFO) << "FFDevice: changing effect failed";
+	} else  {
+        if (write(FFDevice::event_fd, (const void*)&FFDevice::play, sizeof(FFDevice::play)) == -1) {
+            LOG(INFO) << "FFDevice: play effect failed";
+        }
+    }
+}
+
+void FFDevice::off() {
+    if (FFDevice::event_fd == 0) {
+        return;
+    }
+
+    /* Cleanup */
+    if (ioctl(FFDevice::event_fd, EVIOCRMFF, FFDevice::effect.id) == -1) {
+        LOG(INFO) << "FFDevice: remove effect failed";
+    }
+    close(FFDevice::event_fd);
+    FFDevice::event_fd = 0;
+    FFDevice::rumble_effect_id = -1;
+}
+
+bool FFDevice::findEventPath(const char *input_path, char *event_path) {
+    char tmp[256];
+    bool event_exist = false;
+    /* LOG(INFO) << "FFDevice::findEventPath(" << input_path << ")"; */
+	for (int count = 0; count < MAXINPUT; count++) {
+        sprintf(tmp, "%s/input%d", input_path, count);
+        /* LOG(INFO) << "FFDevice: searching in: " << tmp; */
+        if (dir_exists(tmp)) {
```

Comment:
Consider using c++17 and https://en.cppreference.com/w/cpp/filesystem

Reply:
filesystem cannot be used, the current NDK does not support this library :(